### ISSUE TITLE
Documentation of llama.h

### ISSUE THREAD

**Issue by MarcusDunn** [ggerganov/llama.cpp#3870]
I'm writing an [guidance](https://github.com/guidance-ai/guidance) inspired inference server using llama.cpp - I'm much more comfortable in rust and have generated bindings to `llama.h`.

I'm able to make progress by translating examples to rust, testing it, then trial and error'ing my way to the code I want, however I'd like to make a *safe* (in the rust sense) wrapper around llama.cpp then publish and maintain it. 

This involves knowing the implicit invariant used in `llama.h` (both currently and as the project moves forward) - which I don't think I can reasonably do at the moment.

Some examples of things I *think* are true and would like to see documented.
- ~`llama_batch.n_tokens` must never exceed the `n_tokens` used in `llama_batch_init` (basically document what memory is safe to write to in `llama_batch` after init.~ This is already done.
- `get_logits_ith` must be called with an `i` where `batch.logits[i] = true` and then `batch` was decoded. (currently references `lama_eval`)
- All the structs are safe to send between threads. (no thread locals and such)

I understand this project is *very* fast moving and I'd be fine with taking on some of the maintenance burden of keeping the docs up to date. I imagine this would benefit a lot of the others who are maintaining libraries that depend on `llama.h`.

If it is welcome I'll be making small documentation pull requests to this effect, mostly in an effort to get feedback if my assumptions are correct.

Items:
- [ ] `llama_model` and `llama_context`
- [ ] `llama_decode` (only error codes currently documented)
- [ ] `llama_batch.n_seq_id`

**Comment by MarcusDunn** [comment]
think this moreso belongs in discussions.
